rust/noosphere-sphere/src/authority/write.rs

```diff
@@ -0,0 +1,410 @@
+use crate::{internal::SphereContextInternal, HasMutableSphereContext, HasSphereContext};
+
+use super::SphereAuthorityRead;
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use cid::Cid;
+use noosphere_core::{
+    authority::{generate_capability, Authorization, SphereAction, SphereReference},
+    data::{DelegationIpld, Did, Jwt, Link, RevocationIpld},
+    view::SPHERE_LIFETIME,
+};
+use noosphere_storage::{Storage, UcanStore};
+use tokio_stream::StreamExt;
+use ucan::{
+    builder::UcanBuilder,
+    capability::{Capability, Resource, With},
+    crypto::KeyMaterial,
+};
+
+/// Any type which implements [SphereAuthorityWrite] is able to manipulate the
+/// [AuthorityIpld] section of a sphere. This includes authorizing other keys
+/// and revoking prior authorizations.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait SphereAuthorityWrite<K, S>: SphereAuthorityRead<K, S>
+where
+    K: KeyMaterial + Clone + 'static,
+    S: Storage + 'static,
+{
+    /// Authorize another key by its [Did], associating the authorization with a
+    /// provided display name
+    async fn authorize(&mut self, name: &str, identity: &Did) -> Result<Authorization>;
+
+    /// Revoke a previously granted authorization.
+    ///
+    /// Note that correctly revoking an authorization requires signing with the
+    /// root sphere credential, so generally can only be performed on a type
+    /// that implements [SphereAuthorityEscalate].
+    async fn revoke_authorization(&mut self, authorization: &Authorization) -> Result<()>;
+
+    /// Recover authority by revoking all previously delegated authorizations
+    /// and creating a new one that delegates authority to the specified key
+    /// (given by its [Did]).
+    ///
+    /// Note that correctly recovering authority requires signing with the root
+    /// sphere credential, so generally can only be performed on a type that
+    /// implements [SphereAuthorityEscalate]
+    async fn recover_authority(&mut self, new_owner: &Did) -> Result<Authorization>;
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<C, K, S> SphereAuthorityWrite<K, S> for C
+where
+    C: HasSphereContext<K, S> + HasMutableSphereContext<K, S>,
+    K: KeyMaterial + Clone + 'static,
+    S: Storage + 'static,
+{
+    // TODO: We allow optional human-readable names for authorizations, but this
+    // will bear the consequence of leaking personal information about the user
+    // (e.g., a list of their authorized devices). We should encrypt these
+    // names so that they are only readable by the user themselves.
+    async fn authorize(&mut self, name: &str, identity: &Did) -> Result<Authorization> {
```

Comment:
Took me a bit to get this: `authorize()` authorizes an identity with the `Authorize` ability, but grammatically, any ability is authorized in a sense. The possibility of `authorize` as a generic verb applicable to all abilities, overlapping with an Authorize ability, and being similar-but-related to access-checks (`Access::ReadWrite`) is a bit confusing. No concrete suggestions, still wrapping my head around all of it